test/test_torch.py

```diff
@@ -18083,6 +18083,93 @@ def test_large_linspace(self, device, dtype):
         x = torch.linspace(start, end, steps, dtype=dtype, device=device)
         self.assertGreater(x[1] - x[0], (end - start) / steps)
 
+    def _test_where_scalar_template(self, device, dtype, exec_fn):
+        for with_extremal in [True, False]:
+            for ndims in range(0, 4):
+                shape = self._rand_shape(ndims, min_size=5, max_size=10)
+                for n in range(ndims + 1):
+                    for c in combinations(list(range(ndims)), n):
+                        for scalar_type in [int, float, complex]:
+                            if dtype.is_complex:
+                                condition = self._generate_input(shape, dtype, device, with_extremal).abs() > 0.5
+                            else:
+                                condition = self._generate_input(shape, dtype, device, with_extremal) > 0.5
+
+                            x = self._generate_input(shape, dtype, device, with_extremal)
+
+                            if not dtype.is_complex and scalar_type == complex:
+                                continue
+
+                            scalar_1 = scalar_type(random.random())
+
+                            exec_fn(scalar_type, dtype, condition, x, scalar_1)
+
+    # For current implementation,
+    # below are the valid `TensorDtype` and `ScalarType` combinations.
+    def _where_valid_scalar_tensor_combination(self, scalar_type, dtype):
+        if (scalar_type == int and dtype == torch.long):
+            return True
+        elif (scalar_type == float and dtype == torch.double):
+            return True
+        elif (scalar_type == complex and dtype == torch.complex128):
+            return True
+        return False
+
+    @onlyOnCPUAndCUDA
+    @dtypes(*(torch.testing.get_all_int_dtypes() + torch.testing.get_all_fp_dtypes(include_bfloat16=False) +
+              torch.testing.get_all_complex_dtypes()))
+    def test_where_scalar_invalid_combination_raises(self, device, dtype):
```

Comment:
These tests look pretty good and thorough but I think they can be simplified. What about something like...

```
@dtypes(torch.long, torch.double, torch.cdouble)
def test_where_scalar(self, device, dtype):
  shape = self._rand_shape(3, min_size=5, max_size=10)
  condition = torch.randn(shape, device=device) > .5
  input_types = product('scalar', 'tensor')
  for left_type, right_type in input_types:
    left = torch.randn(1, device=device, dtype=dtype) if left_type == 'scalar' else torch.randn_like(condition, dtype=dtype)
    right = (same as above)
    expected = torch.where(condition, left, right)
    actual = torch.where(condition, (left if left.numel() > 1 else left.item()), (right if right.numel() > 1 else right.item()))
   self.assertEqual(actual, expected)
```

?

I think that and a test that invalid combinations raise would be OK? For that test I'd just write a few manual cases. Like create a condition and three scalars (one int, one float, one complex (`complex(.5, -5)`)) and then try int32, float32, and complex64 input tensors with each of the the scalars and assert the error. Something like:

```
for tensor, scalar in inputs:
  self.assertRaises(RuntimeError):
    torch.where(condition, tensor, scalar)
```

It won't cover every case but it should be sufficient.

What are your thoughts on those recommendations?

Reply:
I think after type promotion it would be useful to have these comprehensive tests. Let me know if you still feel that these are too complex. Will try to simplify them as above.
Thanks!

Reply:
I can believe that. OK. Let's stick with the more thorough tests.